### ISSUE TITLE
Expand ArchUnit Testing

### ISSUE THREAD

**Issue by nrmancuso** [checkstyle/checkstyle#12099]
While ArchUnit was introduced in https://github.com/checkstyle/checkstyle/issues/7469, we only did very basic initial integration (see https://github.com/checkstyle/checkstyle/pull/8112). We would like to expand our usage of ArchUnit.

It would be a good first step to see if we can complete the work started at https://github.com/checkstyle/checkstyle/pull/8223, without having ArchUnit add generated files to the project.

References:
https://blog.jdriven.com/2018/10/testing-the-architecture-archunit-in-practice/
https://www.mobileit.cz/Blog/Pages/arch-unit-1.aspx
https://www.mobileit.cz/Blog/Pages/arch-unit-2.aspx
https://www.mscharhag.com/software-development/architecture-validation-with-archunit

Good "alternative" uses:
https://nexocode.com/blog/posts/archunit-forget-architecture-it-s-a-flexible-and-intelligent-linter/

Let's use this issue to discuss usage ideas.

**Comment by nick-mancuso** [comment]
Here might be a good candidate for ArchUnit, if we decide to pursue it:
https://github.com/checkstyle/checkstyle/issues/12100#issuecomment-1225810498

Forbidding mutable test fields.

**Comment by rnveach** [comment]
I would consider a field immutable only if it's value never changes throughout the life of the test run, including the place where it got its value from. If I can get 2 different values from the same method call when executing it at the start of the test run and any time between then and the end of the test run, then the field isn't immutable.

**Comment by Vyom-Yadav** [comment]
> Forbidding mutable test fields.

Using ArchUnit, we can only make sure that fields are `final`, i.e. something like-
```java
@ArchTest                                                                
public static final ArchRule NEW_RULE = fields().that()                   
    .areDeclaredInClassesThat()                                           
    .haveSimpleNameEndingWith("Test")                                     
    .should()                                                             
    .beFinal();                                                          
```

**Comment by Vyom-Yadav** [comment]
> It would be a good first step to see if we can complete the work started at https://github.com/checkstyle/checkstyle/pull/8223, without having ArchUnit add generated files to the project.

The added files by ArchUnit are due to https://www.archunit.org/userguide/html/000_Index.html#_freezing_arch_rules, if we are looking to integrate it, this can act as a suppression file as it will contain all the violations, another method can be to create custom slices without some classes, ignoring those classes.

**Comment by Vyom-Yadav** [comment]
@nick-mancuso @rnveach Let's approve this.

**Comment by Vyom-Yadav** [comment]
### `ArchRule`s to be implemented:

1. **Partial Immutability check**: This `ArchRule` is divided into 2 parts.
    1. **Util classes immutability check**: This check makes sure that all fields defined in Util classes are `final` and `static` and immutable. If the field contains a generic type, it is checked that the generic type is immutable.
    2. **Stateless Module check**: This check makes sure that all fields defined in stateless modules that are not properties are `final` and are immutable. Also every field other than message keys should be `private`. If the field contains a generic type, it is checked that the generic type is immutable. This also makes sure that modules with immutable fields are annotated with [`@StatelessCheck`](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/StatelessCheck.java) and modules with mutable fields are annotated with [`@FileStatefulCheck`](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/FileStatefulCheck.java) or [`@GlobalStatefulCheck`](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/GlobalStatefulCheck.java).
 
These `ArchRule`s cannot check reference immutability.

2. **Disallow classes in api package to depend on classes in util package**
3. **Validation of test/input package/file structure for AST regression**

**Comment by Vyom-Yadav** [comment]
PR for `Partial Immutability check`: #12135

**Comment by Vyom-Yadav** [comment]
PR for `Disallow classes in api package to depend on classes in util package`: #12149

**Comment by nrmancuso** [comment]
>PR for Disallow classes in api package to depend on classes in util package: https://github.com/checkstyle/checkstyle/pull/12149

@Vyom-Yadav I tried a few different approaches for this test, as the current suppression model in #12149 is a bit vague, and would allow for extension of util methods usage within classes that depend on them without violation. Take a look at [this](https://github.com/checkstyle/checkstyle/commit/d646e2b5ddbe0f111f79eae5405f63cba6a0e782) and share your thoughts. This approach also allows to to use Junit assertions, for added logic around checks.

**Comment by Vyom-Yadav** [comment]
@nrmancuso [This](https://github.com/checkstyle/checkstyle/commit/d646e2b5ddbe0f111f79eae5405f63cba6a0e782) approach is clearly better, and allows us for finer suppressions, let's use this.

**Comment by nrmancuso** [comment]
@Vyom-Yadav see https://github.com/checkstyle/checkstyle/commit/fdd2aac85357f46bd91945dce7000991c2b560fb

**Comment by Vyom-Yadav** [comment]
PR for `Checks are extending only AbstractCheck or AbstractFileSetCheck`: https://github.com/checkstyle/checkstyle/pull/12170

**Comment by Vyom-Yadav** [comment]
PR for `All checkstyle packages are free of cycles`: https://github.com/checkstyle/checkstyle/pull/12190

**Comment by romani** [comment]
all changes are merged, if something else required, open new issue,

**Comment by rnveach** [comment]
2 items were not done at https://github.com/checkstyle/checkstyle/issues/12099#issuecomment-1233749876

**Comment by romani** [comment]
>Checks have all private fields.

I pretty sure it is covered by our modifier Check.

>. No check subpackage should depend on each other.

It is covered by jarch cycle validation

**Comment by rnveach** [comment]
> Checks have all private fields.
> I pretty sure it is covered by our modifier Check.

Which one https://checkstyle.org/config_modifier.html ?

> No check subpackage should depend on each other.
> It is covered by jarch cycle validation

Cycle is 2 way. Package 1 > Package 2 > Package 1. This was 1 way. Package 1 > Package 2. It was to represent checks don't rely on other checks.

**Comment by romani** [comment]
https://checkstyle.org/config_design.html#VisibilityModifier
